### ISSUE TITLE
run, hide and watch URL params missing in the new editor

### ISSUE THREAD

**Issue by khrj** [hackclub/sprig#908]
v1 PR: https://github.com/hackclub/sprig/pull/644/files

>- `run=true` can be set to run the code given in file immediately 
>- `hide=true` can be set to not show the editor, just the run window
>- `watch=X` can be set to refresh the file param every X milliseconds
> Each of these params require `file=` to be set
> 
> Example:
> ```
> https://editor.sprig.hackclub.com/?file=http://localhost:3000/build.js&run=true&hide=true&watch=1000
> ```
> 
> This will reload the file from http://localhost:3000/build.js every second (if it has changed), while also immediately rerunning it, and not displaying the editor

I'm not familiar with the new codebase yet, but I might add these myself later if no one picks it up

**Comment by khrj** [comment]
Use case: https://github.com/khrj/hackcraft#development
These params can together be used to get a live-reload of sorts when working locally (which gives multi-file and TS support)

**Comment by kognise** [comment]
I would be open to making some sort of game embed page that only features a screen without any editor or chrome. As it stands this issue is probably a bit out of scope?

**Comment by khrj** [comment]
That's fair — and I think that should solve the use case. Run from file works, and maybe it can run by default and also have a sane watch param (about 1 second) by default. Also — is the old editor still up somewhere?

**Comment by khrj** [comment]
As for being out of scope I'm not really sure. Honestly I feel like it'd be far easier to integrate into the existing editor? But perhaps I'm wrong, I have no idea how the codebase is structured

**Comment by leomcelroy** [comment]
It's intentional we don't support multi file projects out of the box. We want the submitted games to be self-contained and useful for beginners to read, so we chose not to support tools for developing large projects. The way @khrj set it up is a good solution. If you want to make a more involved Sprig project (which is great) migrate to your own ide. A simple file import can be the only needed escape hatch.

**Comment by khrj** [comment]
Yep!

To clarify,

> However, if I already have a fully set up editor, do I really have a reason to replicate it here for almost no benefit?

I'm not referring to myself when I say "I", just any user who already had a development environment set up offline.

And I agree, official support, especially for multiple files if not TypeScript (although auto-fill based on the global.d.ts file is a huge plus) would be great to have!

**Comment by grymmy** [comment]
Okay, I think we have all the info we need here to promote this to the work queue, thanks @khrj @kognise @leomcelroy

**Comment by khrj** [comment]
> It's intentional we don't support multi file projects out of the box. We want the submitted games to be self-contained and useful for beginners to read, so we chose not to support tools for developing large projects. The way @khrj set it up is a good solution. If you want to make a more involved Sprig project (which is great) migrate to your own ide. A simple file import can be the only needed escape hatch.

Alright then, if more complex toolkits are out-of-scope for the web IDE (which I get the reasoning behind), would it be possible to revive the old params? 

Unless anyone can think of a better solution that allows the same features as that workflow (namely: fully custom pipeline + watch-to-run + full-screen), I think that's rather simple to implement

Making those params discoverable and usable via the UI is definitely an improvement, though

**Comment by grymmy** [comment]
> Alright then, if more complex toolkits are out-of-scope for the web IDE (which I get the reasoning behind), would it be possible to revive the old params?

Yes @khrj - That is the plan RE: this issue.  I have also increased the priority of this issue as it is a regression.

**Comment by kognise** [comment]
@khrj and all—

I'm kinda against adding the params as-is, primarily because it feels like it would add unpredictable state to the editor, and partially because it feels like not the ideal solution. But I understand this is a regression, so I'm down to add a hidden page somewhere that has this behavior so you don't have regressions.

I DO HAVE A HOT TAKE THOUGH! I think a proper dev environment for Sprig would be awesome. The cleanest way to do this is just to make your own wrapper page with this behavior, live reloading, etc.— and while I was working on V2 made the whole engine properly modular and TypeScript-y to enable this! It would also be awesome for the engine to be on NPM which would help greatly with this. Heck, even without this you could pull in Sprig as a submodule. So.... what if we just made a proper wrapper! I would be more than happy to help or even write the majority of it. And I would feel a lot happier with that as the cleaner solution.

> Also — is the old editor still up somewhere?

Check out the [ui-v1-backup](https://github.com/hackclub/sprig/tree/0.2.4-ui-backup) tag!

**Comment by grymmy** [comment]
@kognise suggest adding a new issue describing your proposal, as it seems you have some of the details already in-mind.  I'm not philosophically opposed to this, and I think it would be cool to better support more complex game development in Sprig, but @leomcelroy seems to have some strong opinions here and should ultimately be the go/no-go decision-maker.

**Comment by leomcelroy** [comment]
I broke out Sprig as separate library months ago upon request by people to use it in game jams, but never finished publishing it to NPM because I was unsure how to structure the repos. I'm very supportive of having the game library published. The main thing that is lost is the asset editors when working outside our editor, but one can always use them to draw assets and then copy them into their local IDE. As a short term solution let's make a page with the behavior @khrj is requesting. Lexi let's talk separately about publishing the Sprig engine.
